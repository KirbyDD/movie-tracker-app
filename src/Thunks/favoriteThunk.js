import { favoriteMovie, addFavMovie } from '../actions/index'

export const addFavorite = (body) => {
  return async dispatch => {
    const url = 'http://localhost:3000/api/users/favorites/new';
    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(body)
    }
    try {
      const response = await fetch(url, options)
      if(!response.ok) {
        console.log('Error')
      }
      const newFavorite = await response.json()
      dispatch(addFavMovie(newFavorite))
    }
    catch {

    }
  }
}