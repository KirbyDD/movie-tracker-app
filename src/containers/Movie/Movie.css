.movie {
  display: flex;
  justify-content: space-between;
}

.movie-info-card {
  color: #fff;
  margin-top: 50px;
  width: 600px;
  margin-left: auto;
  margin-right: auto;
  height: 700px;
  position: relative;
}

.movie-poster {
  height: 100vh;
  width: auto;
}

.add-fav-btn,
.movie-login-link {
  width: 180px;
  height: 20px;
  border: none;
  background: rgba(255, 255, 255, 0.452);
  border-radius: 3px;
  cursor: pointer;
  outline: none;
}

.add-fav-btn:hover {
  background: #01D277;
}

.movie-detail-container {
  background-image: linear-gradient( rgb(0, 0, 0), rgba(0, 0, 0, 0));
  width: 100%;
}

.movie-backdrop {
  position: absolute;
  top: 0;
  left: 0;
  z-index: -2;
  width: 100vw;
  height: 100vh;
}

.movie-title {
  font-size: 36px;
  margin-bottom: 50px;
}


.content-p {
  color: rgba(255, 255, 255, 0.452);
}

.content-data {
  color: rgba(255, 255, 255, 0.945);
}

.date-rating-container {
  display: flex;
  justify-content: space-between;
  position: absolute;
  bottom: 100px;
  width: 600px;
}