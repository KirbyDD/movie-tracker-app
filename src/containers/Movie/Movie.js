import React, {Component} from 'react';
import { connect } from 'react-redux';
import './Movie.css'
import { addFavorite, getFavorites, removeFavorite } from '..//../Thunks/favoriteThunk'
import { NavLink } from 'react-router-dom'
export class Movie extends Component {
  constructor({ movie, currentUser, handleClick, getFavorites, favoriteMovies, favorites, removeFavorite }){
    super({ movie, currentUser, handleClick, getFavorites, favoriteMovies, favorites, removeFavorite });
    this.url = 'https://image.tmdb.org/t/p/w500'
  }

  checkIfLoggedIn = () => {
    if(this.props.currentUser.loggedIn) {
      const favMovie = {
        movie_id: this.props.movie.movie_id,
        user_id: this.props.currentUser.userDetail.id,
        title: this.props.movie.title,
        poster_path: this.props.movie.poster_path,
        release_date: this.props.movie.release_date,
        vote_average: this.props.movie.vote_average,
        overview: this.props.movie.overview
      }
      if(!this.checkForFavoritedMovie(this.props.movie)) {
        this.props.handleClick(favMovie, this.props.currentUser.userDetail.id);
        this.props.getFavorites(this.props.currentUser.userDetail.id)
      }
    }
  }
  checkForFavoritedMovie = movie => {
    const result = this.props.favorites[0].find(favMovie => favMovie.movie_id === movie.id)
    if(result !== undefined) {
      return true
    } else {
      return false
    }
  }


  render() {
    return (
      <article className='movie'>
        <div className="movie-detail-container">
          <div className='movie-info-card'>
            <h2 className="movie-title">{this.props.movie.title}</h2>
            <p>{this.props.movie.overview}</p>
          <div className="date-rating-container">
            <p className="content-p">Rating:<span className="content-data"> {this.props.movie.vote_average} / 10</span></p>
            <p className="content-p">Release Date:<span className="content-data"> {this.props.movie.release_date}</span></p>
            {this.props.currentUser.loggedIn && !this.checkForFavoritedMovie(this.props.movie) &&
            <button className="add-fav-btn" onClick={e => this.checkIfLoggedIn()}>Add to favorites</button> }
            {this.props.currentUser.loggedIn && this.checkForFavoritedMovie(this.props.movie) && 
            <button onClick={() => this.props.removeFavorite(this.props.currentUser.userDetail.id, this.props.movie.id)} className="movie-login-link">Remove from favorites</button>}
            {!this.props.currentUser.loggedIn && <NavLink  to="/login" className="movie-login-link">Sign in to Favorite</NavLink>}
          </div>
          </div>

          <img src={this.url+this.props.movie.backdrop_path} alt='Movie poster' className="movie-backdrop"/>
        </div>
      <img src={this.url+this.props.movie.poster_path} alt='Movie poster' className="movie-poster"/>
      </article>
    )
    }
}


export const mapStateToProps = state => ({
  currentUser: state.currentUser,
  favorites: state.favorites
})

export const mapDispatchToProps = dispatch => ({
  handleClick: (favMovie, userId) => dispatch(addFavorite(favMovie, userId)),
  getFavorites: (favoriteMovies) => dispatch(getFavorites(favoriteMovies)),
  removeFavorite: (userId, movieId) => dispatch(removeFavorite(userId, movieId))
})

export default connect(mapStateToProps, mapDispatchToProps)(Movie);