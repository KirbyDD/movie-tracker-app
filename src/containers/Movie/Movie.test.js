import React from 'react';
import {Movie, mapStateToProps, mapDispatchToProps} from './Movie';
import { addFavorite, getFavorites, removeFavorite} from '..//../Thunks/favoriteThunk'
import {shallow, mount} from 'enzyme';

let favorites = [
  {
    id: 2,
    title: 'Booboo the Fool',
    poster_path: "/ziEuG1essDuWuC5lpWUaw1uXY2O.jpg",
    release_date: '2019-11-26',
    vote_average: 4.9,
    overview: 'I aint one of your little friends'
  },
  {
    id: 3,
    title: 'BlueFace Baby',
    poster_path: "/ziEuG1essDuWuC5lpWUaw1uXY2O.jpg",
    release_date: '2019-01-26',
    vote_average: 7.9,
    overview: 'Mopped the floor and hid the wet sign just to catch them slipping.'
  }
]

let favoritesTwo = [
  {
    id: 1,
    title: 'The Adventure of TronKat',
    poster_path: "/ziEuG1essDuWuC5lpWUaw1uXY2O.jpg",
    release_date: '2019-01-26',
    vote_average: 9.9,
    overview: 'TronKat is the worlds greatest gamer. But when the world needed him most he vanished.',
    movie_id: 1
  },
  {
    id: 2,
    title: 'Booboo the Fool',
    poster_path: "/ziEuG1essDuWuC5lpWUaw1uXY2O.jpg",
    release_date: '2019-11-26',
    vote_average: 4.9,
    overview: 'I aint one of your little friends',
    movie_id: 2
  },
  {
    id: 3,
    title: 'BlueFace Baby',
    poster_path: "/ziEuG1essDuWuC5lpWUaw1uXY2O.jpg",
    release_date: '2019-01-26',
    vote_average: 7.9,
    overview: 'Mopped the floor and hid the wet sign just to catch them slipping.',
    movie_id: 3
  }
]

const movie = {
  id: 1,
  title: 'The Adventure of TronKat',
  poster_path: "/ziEuG1essDuWuC5lpWUaw1uXY2O.jpg",
  release_date: '2019-01-26',
  vote_average: 9.9,
  overview: 'TronKat is the worlds greatest gamer. But when the world needed him most he vanished.',
  movie_id: 1
}

const currentUser1 = {
  loggedIn: false,
  userDetails: {}
}

const currentUser2 = {
  loggedIn: true,
  userDetail: {
    id: 12
  }
}

describe('Movie', () => {
  let wrapper;
  let handleClick;
  let getFavorites;
  let favMovie;
  let removeFavorite;
  
  beforeEach(function() { 
    handleClick = jest.fn();
    getFavorites = jest.fn();
    removeFavorite = jest.fn();
    favMovie = {
      movie_id: 1,
      user_id: 12,
      title: 'The Adventure of TronKat',
      poster_path: "/ziEuG1essDuWuC5lpWUaw1uXY2O.jpg",
      release_date: '2019-01-26',
      vote_average: 9.9,
      overview: 'TronKat is the worlds greatest gamer. But when the world needed him most he vanished.'
    }
    
    wrapper = shallow(
      <Movie movie={movie}
            currentUser={currentUser2} 
            handleClick={handleClick}
            getFavorites={getFavorites}
            favorites={favorites}
            removeFavorite={removeFavorite}
            />
    )
  })

  it('Should match Logged Out Snapshot', () => {
    wrapper = shallow(
      <Movie movie={movie}
            currentUser={currentUser1} 
            handleClick={handleClick}
            getFavorites={getFavorites}/>
    )
    expect(wrapper).toMatchSnapshot();
  })

  it('Should match Logged In Snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should call checkIfLoggedIn when button is clicked', () => {
    wrapper.instance().checkIfLoggedIn = jest.fn();
    wrapper.find('button').simulate('click');
    expect(wrapper.instance().checkIfLoggedIn).toHaveBeenCalled();
  })

  it('should call two other functions', () => {
    wrapper.instance().checkIfLoggedIn();
    expect(wrapper.instance().props.handleClick).toHaveBeenCalled();
    expect(wrapper.instance().props.getFavorites).toHaveBeenCalled();
  })

  it('handleClick should pass in an object when called', () => {
    wrapper.instance().checkIfLoggedIn();
    expect(wrapper.instance().props.handleClick).toHaveBeenCalledWith(favMovie, 12)
  })

  it('getFavorite should pass in an id when called', () => {
    let id = 12
    wrapper.instance().checkIfLoggedIn();
    expect(wrapper.instance().props.getFavorites).toHaveBeenCalledWith(id);
  })

  it('should have a base url', () => {
    expect(wrapper.instance().url).toEqual('https://image.tmdb.org/t/p/w500')
  })

  it('should call checkIfLoggedIn when button is clicked', () => {
    wrapper = shallow(
      <Movie movie={movie}
            currentUser={currentUser2} 
            handleClick={handleClick}
            getFavorites={getFavorites}
            favorites={favoritesTwo}
            removeFavorite={removeFavorite}
            />
    )
    
    wrapper.find('button').simulate('click');
    expect(wrapper.instance().props.removeFavorite).toHaveBeenCalledWith(12, 1);
  })

  describe('mapDispatchToProps', () => {
    let favMovie = {
      movie_id: 1,
      user_id: 12,
      title: 'The Adventure of TronKat',
      poster_path: "/ziEuG1essDuWuC5lpWUaw1uXY2O.jpg",
      release_date: '2019-01-26',
      vote_average: 9.9,
      overview: 'TronKat is the worlds greatest gamer. But when the world needed him most he vanished.'
    }
    // it('calls dispatch with an addFavorite action when handleSubmit is called', () => {
    //   const mockDispatch = jest.fn();
    //   const actionToDispatch = addFavorite(favMovie, 12);
    //   const mappedProps = mapDispatchToProps(mockDispatch);
    //   mappedProps.handleClick('Something', 12);
    //   expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    // })

    // it('calls dispatch with an addFavorite action when handleSubmit is called', () => {
    //   const mockDispatch = jest.fn();
    //   const actionToDispatch = getFavorites(favoritesTwo);
    //   const mappedProps = mapDispatchToProps(mockDispatch);
    //   mappedProps.getFavorites(favoritesTwo);
    //   expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    // })
  })
})