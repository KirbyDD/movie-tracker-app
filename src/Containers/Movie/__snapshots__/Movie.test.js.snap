// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Should match Logged In Snapshot 1`] = `
<article
  className="movie"
>
  <div
    className="movie-detail-container"
  >
    <div
      className="movie-info-card"
    >
      <h2
        className="movie-title"
      >
        The Adventure of TronKat
      </h2>
      <p>
        TronKat is the worlds greatest gamer. But when the world needed him most he vanished.
      </p>
      <div
        className="date-rating-container"
      >
        <p
          className="content-p"
        >
          Rating:
          <span
            className="content-data"
          >
             
            9.9
             / 10
          </span>
        </p>
        <p
          className="content-p"
        >
          Release Date:
          <span
            className="content-data"
          >
             
            2019-01-26
          </span>
        </p>
        <button
          className="add-fav-btn"
          onClick={[Function]}
        >
          Add to favorites
        </button>
      </div>
    </div>
    <img
      alt="Movie poster"
      className="movie-backdrop"
      src="https://image.tmdb.org/t/p/w500undefined"
    />
  </div>
  <img
    alt="Movie poster"
    className="movie-poster"
    src="https://image.tmdb.org/t/p/w500/ziEuG1essDuWuC5lpWUaw1uXY2O.jpg"
  />
</article>
`;

exports[`Movie Should match Logged Out Snapshot 1`] = `
<article
  className="movie"
>
  <div
    className="movie-detail-container"
  >
    <div
      className="movie-info-card"
    >
      <h2
        className="movie-title"
      >
        The Adventure of TronKat
      </h2>
      <p>
        TronKat is the worlds greatest gamer. But when the world needed him most he vanished.
      </p>
      <div
        className="date-rating-container"
      >
        <p
          className="content-p"
        >
          Rating:
          <span
            className="content-data"
          >
             
            9.9
             / 10
          </span>
        </p>
        <p
          className="content-p"
        >
          Release Date:
          <span
            className="content-data"
          >
             
            2019-01-26
          </span>
        </p>
        <NavLink
          className="movie-login-link"
          to="/login"
        >
          Sign in to Favorite
        </NavLink>
      </div>
    </div>
    <img
      alt="Movie poster"
      className="movie-backdrop"
      src="https://image.tmdb.org/t/p/w500undefined"
    />
  </div>
  <img
    alt="Movie poster"
    className="movie-poster"
    src="https://image.tmdb.org/t/p/w500/ziEuG1essDuWuC5lpWUaw1uXY2O.jpg"
  />
</article>
`;
